rust/src/treefile.rs

```diff
@@ -1306,24 +1306,21 @@ impl TreeComposeConfig {
 
     // we need to ensure that appended repo packages override earlier ones
     fn handle_repo_packages_overrides(&mut self) {
-        if let Some(repo_packages) = self.repo_packages.take() {
+        self.repo_packages.as_mut().map(|repo_packages| {
             let mut seen_pkgs: HashSet<String> = HashSet::new();
-            self.repo_packages = Some({
-                let mut v: Vec<RepoPackage> = repo_packages
-                    .into_iter()
-                    .rev()
-                    .map(|mut rp| {
-                        rp.packages.retain(|p| seen_pkgs.insert(p.into()));
-                        rp
-                    })
-                    .filter(|rp| !rp.packages.is_empty())
-                    .collect();
-                // can't inline this in the iterator chain above:
-                // https://doc.rust-lang.org/std/iter/struct.Map.html#notes-about-side-effects
-                v.reverse();
-                v
-            });
-        }
+            // Create a temporary new filtered vec
+            let v: Vec<_> = repo_packages
+                .drain(..)
+                .rev()
+                .map(|mut rp| {
+                    rp.packages.retain(|p| seen_pkgs.insert(p.into()));
+                    rp
+                })
+                .filter(|rp| !rp.packages.is_empty())
+                .collect();
+            // Now replace the original, re-reversing
```

Comment:
Let's keep the link to https://doc.rust-lang.org/std/iter/struct.Map.html#notes-about-side-effects? That was a really tricky bit which took me a few to grok.